standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java

```diff
@@ -1396,245 +388,59 @@ private void insertAbortRetryRetentionTimeOnError(Connection dbConn, CompactionI
-          }
-        }
-        if(ci.id == 0) {
-          //The failure occurred before we even made an entry in COMPACTION_QUEUE
-          //generate ID so that we can make an entry in COMPLETED_COMPACTIONS
-          ci.id = generateCompactionQueueId(stmt);
-          //this is not strictly accurate, but 'type' cannot be null.
-          if(ci.type == null) {
-            ci.type = CompactionType.MINOR;
-          }
-          ci.start = getDbTime(dbConn);
-          LOG.debug("The failure occurred before we even made an entry in COMPACTION_QUEUE. Generated ID so that we "
-                  + "can make an entry in COMPLETED_COMPACTIONS. New Id: {}", ci.id);
-        }
-        close(rs, stmt, null);
-        closeStmt(pStmt);
+    long endTime = getDbTime().getTime();
+    if (ciActual != null) {
+      //preserve errorMessage and state
+      ciActual.errorMessage = ci.errorMessage;
+      ciActual.state = ci.state;
 
-        pStmt = dbConn.prepareStatement(TxnQueries.INSERT_INTO_COMPLETED_COMPACTION);
-        CompactionInfo.insertIntoCompletedCompactions(pStmt, ci, getDbTime(dbConn));
-        int updCount = pStmt.executeUpdate();
-        LOG.debug("Inserted {} entries into COMPLETED_COMPACTIONS", updCount);
-        closeStmt(pStmt);
-        dbConn.commit();
-      } catch (SQLException e) {
-        LOG.error("Failed to mark compaction request as " + strState + ", rolling back transaction: " + ci, e);
-        rollbackDBConn(dbConn);
-        checkRetryable(e, "updateStatus(" + ci + ")");
-      } finally {
-        close(rs, stmt, null);
-        close(null, pStmt, dbConn);
+      dataSourceWrapper.getJdbcTemplate().update("DELETE FROM \"COMPACTION_QUEUE\" WHERE \"CQ_ID\" = :id",
+          new MapSqlParameterSource("id", ci.id));
+    } else {
+      if (ci.id > 0) {
+        //the record with valid CQ_ID has disappeared - this is a sign of something wrong
+        throw new IllegalStateException("No record with CQ_ID=" + ci.id + " found in COMPACTION_QUEUE");
+      }
+      ciActual = ci;
+    }
+    if (ciActual.id == 0) {
+      //The failure occurred before we even made an entry in COMPACTION_QUEUE
+      //generate ID so that we can make an entry in COMPLETED_COMPACTIONS
+      ciActual.id = generateCompactionQueueId();
+      //this is not strictly accurate, but 'type' cannot be null.
+      if (ciActual.type == null) {
+        ciActual.type = CompactionType.MINOR;
       }
-    } catch (RetryException e) {
-      updateStatus(ci);
+      //in case of creating a new entry start and end time will be the same
+      ciActual.start = endTime;
+      LOG.debug("The failure occurred before we even made an entry in COMPACTION_QUEUE. Generated ID so that we "
+          + "can make an entry in COMPLETED_COMPACTIONS. New Id: {}", ciActual.id);
     }
+
+    new InsertCompactionInfoCommand(ciActual, endTime).execute(dataSourceWrapper);
```

Comment:
not consistent naming, in some places handler in other command

Reply:
It is consistent. There are two base classes: `QueryHandler` and `ParameterizedCommand`. The former is for queries (with `ResultSet` mapping), the latter is for INSERT/UPDATE/DELETE/etc commands.